File: src/SkillTrail.Biz/Entites/Progress.cs
using SkillTrail.Biz.Abstractions;
using System.ComponentModel.DataAnnotations.Schema;

namespace SkillTrail.Biz.Entites
{
    public sealed class Progress : EntityBase
    {
        public string TaskId { get; set; } = string.Empty;
        public string UserId { get; set; } = string.Empty;
        public ProgressStatus Status { get; set; } = ProgressStatus.None;
        public string Note { get; set; } = string.Empty;

        [ForeignKey(nameof(TaskId))]
        public Task Task { get; set; } = null!;

        [ForeignKey(nameof(UserId))]
        public User User { get; set; } = null!;
    }

    public enum ProgressStatus
    {
        None,
        NotStarted,
        InProgress,
        Completed,
    }
}

File: src/skilltrail.client/src/entities/progress/model/progress.ts
import type dayjs from "dayjs";

export const progressStatus = {
    NONE: 0,
    NOT_STARTED: 1,
    IN_PROGRESS: 2,
    COMPLETED: 3,
} as const;

export type ProgressStatus = typeof progressStatus[keyof typeof progressStatus];

export interface Progress {
    id: string;
    taskId: string;
    userId: string;
    status: ProgressStatus;
    updateDateTime: dayjs.Dayjs;
}

// 進捗クエリサービスモデル（TaskとJoinした結果）
export interface ProgressQueryServiceModel {
    id: string;
    taskId: string;
    level: number;
    taskName: string;
    userId: string;
    status: ProgressStatus;
    note: string;
}

// 進捗更新リクエスト
export interface UpdateProgressRequest {
    taskId: string;
    status: ProgressStatus;
    note: string;
}

// ユーザーID指定での進捗取得リクエスト
export interface GetProgressByUserIdRequest {
    userId: string;
}

// ID指定での進捗取得リクエスト
export interface GetProgressByIdRequest {
    id: string;
}

export type UpdateProgressResponse = {
    prevLevel: number;
    newLevel: number;
}